File: dynamic_programming/linear_dp/no_chain/0343_integer_break.js
/*
Given an integer n, break it into the sum of k positive integers, 
where k >= 2, and maximize the product of those integers.

Return the maximum product you can get.
2 <= n <= 58
*/

// https://leetcode.com/problems/integer-break/description/

// let dp[i] be the maximum product of breaking i
// for every j that, 0 < j < i, we can break i into j and i - j
// we further break j into two parts, if and only if dp[j] > j
// we further break i - j into two parts, if and only if dp[i - j] > i - j
var integerBreak = function (n) {
  let dp = new Array(n + 1).fill(0);
  for (let i = 2; i <= n; i++) {
    // the minimum product of breaking i is 1 * (i - 1)
    dp[i] = 1 * (i - 1);
    let mid = Math.floor(i / 2);
    for (let j = 1; j <= mid; j++) {
      dp[i] = Math.max(dp[i], Math.max(j, dp[j]) * Math.max(i - j, dp[i - j]));
    }
  }
  return dp[n];
};

File: dynamic_programming/linear_dp/no_chain/0343_integer_break.py
# Given an integer n, break it into the sum of k positive integers, 
# where k >= 2, and maximize the product of those integers.

# Return the maximum product you can get.
# 2 <= n <= 58

# https://leetcode.com/problems/integer-break/

# math: for n > 4: make each piece integer the closest to e, which is 3
class Solution:
    def integerBreak(self, n: int) -> int:
        if n == 2: return 1
        if n == 3: return 2
        if n == 4: return 4
        else:
            ans = 1
            while n > 4:
                ans *= 3
                n -= 3
            return ans * n

# dp[i] 表示为：将正整数 i 拆分为至少 2 个正整数的和之后，这些正整数的最大乘积。
class Solution:
    def integerBreak(self, n: int) -> int:
        dp = [0 for _ in range(n + 1)]
        for i in range(2, n + 1):
            for j in range(i):
                dp[i] = max(dp[i], (i - j) * j, dp[i - j] * j)
        return dp[n]
